vortex-array/src/arrays/list/mod.rs

```diff
@@ -100,40 +101,58 @@ impl ListArray {
         })
     }
 
-    // TODO: merge logic with varbin
-    // TODO(ngates): should return a result if it requires canonicalizing offsets
-    pub fn offset_at(&self, index: usize) -> usize {
+    pub fn offset_at(&self, index: usize) -> VortexResult<usize> {
+        if index > self.len() + 1 {
+            vortex_bail!(OutOfBounds: index, 0, self.len() + 1)
+        }
+
         self.offsets()
             .as_opt::<PrimitiveVTable>()
-            .map(|p| match_each_native_ptype!(p.ptype(), |P| { p.as_slice::<P>()[index].as_() }))
+            .map(|p| {
+                Ok(match_each_native_ptype!(p.ptype(), |P| {
+                    p.as_slice::<P>()[index].as_()
+                }))
+            })
             .unwrap_or_else(|| {
                 self.offsets()
                     .scalar_at(index)
-                    .unwrap_or_else(|err| {
-                        vortex_panic!(err, "Failed to get offset at index: {}", index)
-                    })
-                    .as_ref()
-                    .try_into()
-                    .vortex_expect("Failed to convert offset to usize")
+                    .and_then(|s| usize::try_from(&s))
             })
     }
 
-    // TODO: fetches the elements at index
+    /// Returns the elements at the given index from the list array.
     pub fn elements_at(&self, index: usize) -> VortexResult<ArrayRef> {
-        let start = self.offset_at(index);
-        let end = self.offset_at(index + 1);
+        let start = self.offset_at(index)?;
+        let end = self.offset_at(index + 1)?;
+        self.elements().slice(start, end)
+    }
+
+    /// Returns elements of the list array referenced by the offsets array
+    pub fn sliced_elements(&self) -> VortexResult<ArrayRef> {
+        let start = self.offset_at(0)?;
+        let end = self.offset_at(self.len())?;
         self.elements().slice(start, end)
     }
 
-    // TODO: fetches the offsets of the array ignoring validity
+    /// Returns the offsets array.
     pub fn offsets(&self) -> &ArrayRef {
         &self.offsets
     }
 
-    // TODO: fetches the elements of the array ignoring validity
+    /// Returns the elements array.
     pub fn elements(&self) -> &ArrayRef {
         &self.elements
     }
+
+    /// Create a copy of this array by adjusting offsets to start at 0 and removing elements not referenced by the offsets
+    pub fn offset_to_0(&self) -> VortexResult<Self> {
```

Comment:
not convinced by the name. In the end though its changing starting offset to 0 is why you need to prune elements